rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+  void (*block_host_until_done)(
+      SE_Device* executor, SE_Stream stream,
+      TF_Status* status);
+
+  // Synchronizes all activity occurring in the StreamExecutor's context (most
+  // likely a whole device).
+  TF_BOOL (*synchronize_all_activity)(SE_Device* executor);
+
+  // Obtains metadata about the underlying device.
+  void (*fill_device_description)(SE_Device* executor,
+                               SE_DeviceDescription* description,
+                               TF_Status* status);
+
+  // Entrains on a stream a user-specified function to be run on the host.
+  TF_BOOL (*host_callback)(SE_Device* executor, SE_Stream* stream,
+                     TF_StatusCallbackFn callback_fn, void* ctx);
+} SE_StreamExecutor;
+
+#define SE_STREAMEXECUTOR_STRUCT_SIZE TF_OFFSET_OF_END(SE_StreamExecutor, host_callback)
+
+typedef struct SE_PlatformParams {
+  size_t struct_size;
+  SE_PlatformId* id;
+  
+  // Callbacks for creating/destroying SE_Device.
+  SE_Device* (*create_device)(SE_Options* options, TF_Status* status);
+  void (*destroy_device)(SE_Device* device);
+  
+  // Callbacks for creating/destroying SE_StreamExecutor.
+  SE_StreamExecutor* (*create_stream_executor)(TF_Status* status);
+  void (*destroy_stream_executor)(SE_StreamExecutor* stream_executor);
+} SE_PlatformParams;
+
+#define SE_PLATFORM_PARAMS_SIZE TF_OFFSET_OF_END(SE_PlatformParams, destroy_stream_executor)
+
+TF_CAPI_EXPORT SE_Platform* SE_NewPlatform(SE_PlatformParams params);
+
+typedef struct SE_PlatformRegistartionParams {
+  size_t struct_size;
+  
+  // Platform name
+  const char* name;
+  size_t name_len;
+  
+  // Device type name. Right now only GPU is supported.
+  char* type;
+  size_t type_len;
+  
+  SE_Platform* platform;
+} SE_PlatformParams;
+
+#define SE_PLATFORM_REGISTRATION_PARAMS_SIZE TF_OFFSET_OF_END(SE_PlatformRegistrationParams, platform)
+
+TF_CAPI_EXPORT void SE_RegisterPlatform(
+     SE_PlatformRegistartionParams params,
+     TF_Status* status);
+
+#ifdef __cplusplus
+} // extern "C"
+#endif
+```
+
+## Registration implementation
+
+Registration will be implemented by registering a new StreamExecutor platform as well as a new TensorFlow device with [DeviceFactory](https://osscs.corp.google.com/tensorflow/tensorflow/+/master:tensorflow/core/common_runtime/device_factory.h;l=30?q=DeviceFactory).
+
+```cpp
+void SE_RegisterPlatform(
+    SE_PlatformRegistrationParams params, TF_Status* status) {
+   // Register new platform
+   std::unique_ptr<stream_executor::internal::CPlatform> platform(
+        new stream_executor::internal::CPlatform(params));
+   SE_CHECK_OK(
+      stream_executor::MultiPlatformManager::RegisterPlatform(
+          std::move(platform)));
+   
+   // Register PluggableDevice
+   std::string platform_name_str(params.name, params.name_len);
+   std::string type_str(params.type, params.type_len);
+   DeviceFactory::Register(type_str, new PluggableDeviceFactory(platform_name_str), priority);
```

Comment:
does this **priority** also be registered from plugin or it will be defined in tensorflow proper? because I didn't find this priority in `SE_PlatformRegistartionParams` or `SE_RegisterPlatform` parameter. Thanks

Reply:
I am thinking this could be a fixed value we set behind the scenes. Fixed priority value of plugins should be higher than any default devices. At the same time, every plugin should have the same fixed priority value.

Reasoning for fixed value: Suppose plugins can choose priority. Then, plugin provided by 1 vendor might set a higher value than 2nd vendor but may be 2nd vendor wants its plugin to be selected at higher priority. Basically, I don't see a good way that plugin authors would pick priority values relative to other plugins.

Reply:
I see it. A potential issue I see here is if multiple plugins register the same name "GPU" and the same priority, only one plugin can register successful, perhaps platform_name can be a part of key,  but this will need a big change and make things more complicated, such device id unifying.

Reply:
You are right that such issue can come up. A simple way is to register neither plugin if there is a conflict and print errors.

A better approach would be to let users specify platform name they prefer if there is a conflict. That would need to become an option in Python API. Definitely something to consider.